sound/soc/sof/ipc.c

```diff
@@ -612,6 +623,169 @@ int snd_sof_ipc_stream_posn(struct snd_sof_dev *sdev,
 }
 EXPORT_SYMBOL(snd_sof_ipc_stream_posn);
 
+static int sof_get_ctrl_size_params(struct snd_sof_control *scontrol,
+				    u32 ipc_cmd,
+				    enum sof_ipc_ctrl_type ctrl_type,
+				    enum sof_ipc_ctrl_cmd ctrl_cmd,
+				    struct sof_ipc_ctrl_data *cdata,
+				    struct sof_ctrl_data_send_params *sipc)
+{
+	switch (ctrl_type) {
+	case SOF_CTRL_TYPE_VALUE_CHAN_GET:
+	case SOF_CTRL_TYPE_VALUE_CHAN_SET:
+		sipc->msg_bytes = scontrol->num_channels *
+			sizeof(struct sof_ipc_ctrl_value_chan);
+		sipc->hdr_bytes = sizeof(struct sof_ipc_ctrl_data);
+		cdata->total_elems = scontrol->num_channels;
+		break;
+	case SOF_CTRL_TYPE_VALUE_COMP_GET:
+	case SOF_CTRL_TYPE_VALUE_COMP_SET:
+		sipc->msg_bytes = scontrol->num_channels *
+			sizeof(struct sof_ipc_ctrl_value_comp);
+		sipc->hdr_bytes = sizeof(struct sof_ipc_ctrl_data);
+		cdata->total_elems = scontrol->num_channels;
+		break;
+	case SOF_CTRL_TYPE_DATA_GET:
+	case SOF_CTRL_TYPE_DATA_SET:
+		sipc->msg_bytes = cdata->data->size;
+		sipc->hdr_bytes = sizeof(struct sof_ipc_ctrl_data) +
+			sizeof(struct sof_abi_hdr);
+		cdata->total_elems = cdata->data->size;
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	cdata->rhdr.hdr.cmd = SOF_IPC_GLB_COMP_MSG | ipc_cmd;
+	cdata->cmd = ctrl_cmd;
+	cdata->type = ctrl_type;
+	cdata->rhdr.hdr.size = sipc->msg_bytes + sipc->hdr_bytes;
+	cdata->comp_id = scontrol->comp_id;
+	cdata->msg_id = 0;
+	cdata->num_elems = cdata->total_elems;
+
+	return 0;
+}
+
+static int sof_get_ctrl_copy_params(enum sof_ipc_ctrl_type ctrl_type,
+				    struct sof_ipc_ctrl_data *cdata,
+				    struct sof_ipc_ctrl_data *hdr,
+				    struct sof_ctrl_data_send_params *sipc)
+{
+	switch (ctrl_type) {
+	case SOF_CTRL_TYPE_VALUE_CHAN_GET:
+	case SOF_CTRL_TYPE_VALUE_CHAN_SET:
+		sipc->src = (unsigned char *)cdata->chanv;
+		sipc->dst = (unsigned char *)hdr->chanv;
+		break;
+	case SOF_CTRL_TYPE_VALUE_COMP_GET:
+	case SOF_CTRL_TYPE_VALUE_COMP_SET:
+		sipc->src = (unsigned char *)cdata->compv;
+		sipc->dst = (unsigned char *)hdr->compv;
+		break;
+	case SOF_CTRL_TYPE_DATA_GET:
+	case SOF_CTRL_TYPE_DATA_SET:
+		sipc->src = (unsigned char *)cdata->data->data;
+		sipc->dst = (unsigned char *)hdr->data->data;
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	/* calculate payload size and number of messages */
+	sipc->pl_size = SOF_IPC_MSG_MAX_SIZE - sipc->hdr_bytes;
+	sipc->num_msg = DIV_ROUND_UP(sipc->msg_bytes, sipc->pl_size);
+	sipc->offset = 0;
+
+	return 0;
+}
+
+static int sof_send_recv_ctrl_data(struct snd_sof_dev *sdev,
+				   struct snd_sof_control *scontrol,
+				   u32 ipc_cmd,
+				   enum sof_ipc_ctrl_type ctrl_type,
+				   enum sof_ipc_ctrl_cmd ctrl_cmd)
+{
+	struct sof_ipc_ctrl_data *cdata = scontrol->control_data;
+	struct sof_ipc_ctrl_data *hdr;
```

Comment:
@juimonen sorry it took me a little longer to review this. But calling this hdr is so confusing. This is infact data just a portion of it. I would go for something more intuitive like "wholedata" and "partdata" to give an example.